include/fc/io/raw.hpp

```diff
@@ -172,9 +172,11 @@ namespace fc {
       uint64_t v = 0; char b = 0; uint8_t by = 0;
       do {
           s.get(b);
+          if( by >= 64 || (by == 63 && b > 1) )
```

Comment:
`b > 1`: should convert `b` to unsigned.

Reply:
Actually, this check can be changed to `if( by == 63 && b != 1 )`, since
* we can reject the data if `b==0` (it doesn't make sense to add one more byte which is zero);
* we will either end the loop or throw an exception when `b` is `63`, so `by >= 64` will always be false thus can be removed.

Reply:
If ``by == 63 && b < 0`` then ``uint8_t(b) & 0x80 != 0``, so the while loop will continue, and in the next round ``by >= 64`` will trigger the condition and throw.

``by >= 64`` cannot be left out, because I removed the check from the while loop (want to throw, not just exit the loop).

We could disallow the case ``by == 63 && b == 0``, but it will be handled by the enclosing logic (deserialize, then re-serialize and check the signature) anyway. Even if we disallow trailing 0 or 0x80 bytes completely, we do not have a complete canonical representation, because we do not know the original type that is being serialized as an unsigned_int here (could be uint8_t, uint16_t, uint32_t, uint64_t).

I think the original intent of the EOS fix was only to avoid the undefined behaviour associated with by growing bigger than the bit size of ``value``.

Reply:
Makes sense, thanks.
I still think it's better to check whether `v < 0` is true here to make the code/logic clearer.

Reply:
Added cast to uint8_t, it's cheaper than separately checking for <0.